src/OpenTelemetry.Collector.Dependencies/Implementation/SqlClientDiagnosticListener.cs

```diff
@@ -0,0 +1,124 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+using System;
+using System.Diagnostics;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Collector.Dependencies.Implementation
+{
+    internal class SqlClientDiagnosticListener : ListenerHandler
+    {
+        private readonly PropertyFetcher commandFetcher = new PropertyFetcher("Command");
+        private readonly PropertyFetcher connectionFetcher = new PropertyFetcher("Connection");
+        private readonly PropertyFetcher dataSourceFetcher = new PropertyFetcher("DataSource");
+        private readonly PropertyFetcher databaseFetcher = new PropertyFetcher("Database");
+        private readonly PropertyFetcher commandTextFetcher = new PropertyFetcher("CommandText");
+        private readonly PropertyFetcher exceptionFetcher = new PropertyFetcher("Exception");
+
+        public SqlClientDiagnosticListener(string sourceName, ITracer tracer)
+            : base(sourceName, tracer)
+        {
+        }
+
+        public void OnCompleted()
+        {
+        }
+
+        public void OnError(Exception error)
+        {
+        }
+
+        public override void OnStartActivity(Activity activity, object payload)
+        {
+        }
+
+        public override void OnCustom(string name, Activity activity, object payload)
+        {
+            switch (name)
+            {
+                case "System.Data.SqlClient.WriteCommandBefore":
+                case "Microsoft.Data.SqlClient.WriteCommandBefore":
+                    {
+                        var command = this.commandFetcher.Fetch(payload);
+
+                        const string EventNameSuffix = ".OnStartActivity";
+                        if (command == null)
+                        {
+                            CollectorEventSource.Log.NullPayload(nameof(SqlClientDiagnosticListener) + EventNameSuffix);
+                            return;
+                        }
+
+                        var connection = this.connectionFetcher.Fetch(command);
+                        var dataSource = this.dataSourceFetcher.Fetch(connection);
+                        var database = this.databaseFetcher.Fetch(connection);
+                        var commandText = this.commandTextFetcher.Fetch(command);
+
+                        this.Tracer.StartActiveSpan(database.ToString(), SpanKind.Client, out var span);
+
+                        span.PutComponentAttribute("SQL Server");
+
+                        span.PutDatabaseTypeAttribute("sql");
+                        span.PutDatabaseInstanceAttribute(database.ToString());
```

Comment:
what is the type of `database` and what `database.ToString()` is expected to return?

Reply:
According to the spec `Database instance name. E.g., In java, if the jdbc.url="jdbc:mysql://db.example.com:3306/customers", the instance name is "customers".`
For example, we have a connection string like that
`Data Source=127.0.0.1;Initial Catalog=test_db;Persist Security Info=True;User ID=admin;Password=admin`
So `db.instance=test_db`

Reply:
so can you explicitly make `database` as string type and don't call ToString?